node/core/chain-selection/src/lib.rs

```diff
@@ -202,96 +209,143 @@ impl Error {
 
-	timestamp_now() + STAGNANT_TIMEOUT
-}
+impl Clock for SystemClock {
+	fn timestamp_now(&self) -> Timestamp {
+		// `SystemTime` is notoriously non-monotonic, so our timers might not work
+		// exactly as expected. Regardless, stagnation is detected on the order of minutes,
+		// and slippage of a few seconds in either direction won't cause any major harm.
+		//
+		// The exact time that a block becomes stagnant in the local node is always expected
+		// to differ from other nodes due to network asynchrony and delays in block propagation.
+		// Non-monotonicity exarcerbates that somewhat, but not meaningfully.
 
-// TODO https://github.com/paritytech/polkadot/issues/3293:
-//
-// This is used just so we can have a public function that calls
-// `run` and eliminates all the unused errors.
-//
-// Should be removed when the real implementation is done.
-struct VoidBackend;
+		match SystemTime::now().duration_since(UNIX_EPOCH) {
+			Ok(d) => d.as_secs(),
+			Err(e) => {
+				tracing::warn!(
+					target: LOG_TARGET,
+					err = ?e,
+					"Current time is before unix epoch. Validation will not work correctly."
+				);
 
-impl Backend for VoidBackend {
-	fn load_block_entry(&self, _: &Hash) -> Result<Option<BlockEntry>, Error> {
-		Ok(None)
-	}
-	fn load_leaves(&self) -> Result<LeafEntrySet, Error> {
-		Ok(LeafEntrySet::default())
-	}
-	fn load_stagnant_at(&self, _: Timestamp) -> Result<Vec<Hash>, Error> {
-		Ok(Vec::new())
-	}
-	fn load_stagnant_at_up_to(&self, _: Timestamp)
-		-> Result<Vec<(Timestamp, Vec<Hash>)>, Error>
-	{
-		Ok(Vec::new())
+				0
+			}
+		}
 	}
-	fn load_first_block_number(&self) -> Result<Option<BlockNumber>, Error> {
-		Ok(None)
+}
+
+/// The interval, in seconds to check for stagnant blocks.
+#[derive(Debug, Clone)]
+pub struct StagnantCheckInterval(Duration);
+
+impl Default for StagnantCheckInterval {
+	fn default() -> Self {
+		// 5 seconds is a reasonable balance between avoiding DB reads and
+		// ensuring validators are generally in agreement on stagnant blocks.
+		//
+		// Assuming a network delay of D, the longest difference in view possible
+		// between 2 validators is D + 5s.
+		const DEFAULT_STAGNANT_CHECK_INTERVAL: Duration = Duration::from_secs(5);
+
+		StagnantCheckInterval(DEFAULT_STAGNANT_CHECK_INTERVAL)
 	}
-	fn load_blocks_by_number(&self, _: BlockNumber) -> Result<Vec<Hash>, Error> {
-		Ok(Vec::new())
+}
+
+impl StagnantCheckInterval {
+	/// Create a new stagnant-check interval wrapping the given duration.
+	pub fn new(interval: Duration) -> Self {
+		StagnantCheckInterval(interval)
 	}
 
-	fn write<I>(&mut self, _: I) -> Result<(), Error>
-		where I: IntoIterator<Item = BackendWriteOp>
-	{
-		Ok(())
+	fn timeout_stream(&self) -> impl Stream<Item = ()> {
+		let interval = self.0;
+		let mut delay = futures_timer::Delay::new(interval);
+
+		futures::stream::poll_fn(move |cx| {
+			let poll = delay.poll_unpin(cx);
+			if poll.is_ready() {
+				delay.reset(interval)
+			}
+
+			poll.map(Some)
+		})
```

Comment:
what about reusing https://github.com/paritytech/polkadot/blob/f560edf6b40e18722d16512cef0e1e0a78eae08c/node/core/pvf/src/host.rs#L601-L609

which I think does the same, but is more readable

Reply:
What we're really missing here is a `futures::stream::poll_fn_async`. I don't think that DRY is accurate in this case; a 'pulse_every' floating in a random util crate is worse than a small amount of code reuse.